src/fastpath/fastpath.c

```diff
@@ -6,15 +6,313 @@
+     * have a fault while invoking the fastpath */
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        slowpath(SysSend);
+    }
+
+    /* Lookup the cap */
+    cap_t cap = lookup_fp(TCB_PTR_CTE_PTR(NODE_STATE(ksCurThread), tcbCTable)->cap, cptr);
+
+    /* Check it's a notification */
+    if (unlikely(!cap_capType_equals(cap, cap_notification_cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Check that we are allowed to send to this cap */
+    if (unlikely(!cap_notification_cap_get_capNtfnCanSend(cap))) {
+        slowpath(SysSend);
+    }
+
+    /* Get the notification address */
+    notification_t *ntfnPtr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(cap));
+
+    /* Get the notification state */
+    uint32_t ntfnState = notification_ptr_get_state(ntfnPtr);
+
+    /* Get the notification badge */
+    word_t badge = cap_notification_cap_get_capNtfnBadge(cap);
+    switch (ntfnState) {
+    case NtfnState_Active: {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfnPtr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfnPtr));
+        restore_user_context();
+        UNREACHABLE();
+    }
+    case NtfnState_Idle: {
+        tcb_t *dest = (tcb_t *) notification_ptr_get_ntfnBoundTCB(ntfnPtr);
+
+        /* Check if we are bound and that thread is waiting for a message */
+        if (dest && thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_BlockedOnReceive) {
+            /* Basically equivalent to maybeDonateSchedContext. Check whether the thread already has
+             * a SC or if one can be donated from the notification. If neither is true, go to
+             * slowpath */
+            if (!dest->tcbSchedContext) {
+                sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfnPtr));
+                if (sc == NULL || sc->scTcb != NULL) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+
+                /* Slowpath FPU saving case. We check that dest either isn't currently using the FPU
+                 * so that we do not have to save its state for migration (lazy FPU) or that it is the
+                 * the thread that is currently using the FPU, but the CPU is the same as that of the
+                 * scheduling context so that tcb migration is not needed. Not exactly the same as the
+                 * slowpath as this saves FPU state in the second case but there should be no reason
+                 * to touch the FPU if migration is not required */
+#ifdef ENABLE_SMP_SUPPORT
+#ifdef CONFIG_HAVE_FPU
+                if (nativeThreadUsingFPU(dest)) {
+                    slowpath(SysSend);
+                    UNREACHABLE();
+                }
+#endif
+#endif
+            } else {
+                sc = dest->tcbSchedContext;
+            }
+
+            /* Signal to higher prio thread is NOT fastpathed. Not sure if this handles idle thread */
```

Comment:
Neither `ksCurThread` or `dest` can be the idle thread. I think the part about the idle thread can be removed from this comment.

Reply:
Oh right, I don't think I was considering the case where `NODE_STATE_ON_CORE(ksCurThread, sc->scCore)` was selecting the current thread on a different core (which could be the idle thread).

Reply:
The comment could be updated to explain that this means, that even if the other core is idling, the slowpath will be taken to signal the destination thread.

Reply:
Okay, so I won't merge in https://github.com/alwin-joshy/seL4/commit/565e043164443b643136eb19e22e3f3ad2b38500?, at least for now